learn/models/conformal-regression/index.qmd

```diff
@@ -0,0 +1,437 @@
+
+test_split_res %>% 
+  ggplot(aes(x)) +
+  geom_point(aes(y = y), alpha = 1 / 20) +
+  geom_line(
+    data = nnet_pred, aes(y = .pred),
+    linewidth = 1, col = "blue"
+  ) +
+  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
+              col = "#D95F02", linewidth = 3 / 4, fill = NA)
+
+cal_resid_width <- 
+  quantile(test_split_res$y - test_split_res$.pred, probs = c(0.05, 0.95)) %>% 
+  diff() %>% 
+  signif(3)
+```
+
+Since we know the outcome values, we can compute the coverage for this particular data set. Since we created a 90% prediction interval, about 90% of the outcomes should be within our bounds. Let's create a function and apply it to these data:
+
+```{r}
+#| label: split-coverage
+coverage <- function(x) {
+  x %>% 
+    mutate(in_bound = .pred_lower < y & .pred_upper > y) %>% 
+    summarise(coverage = mean(in_bound) * 100)
+}
+coverage(test_split_res)
+```
+
+## Using Resampling Results
+
+In a way, the calibration set serves a similar role as a traditional validation set. Since we can just re-predict our training set, we need to evaluate our model on a separate collection of labeled data. 
+
+However, we also know that resampling methods can serve the same purpose. During resampling, we can compute a set of predictions that were not used to fit the model. For example, in 10-fold cross-validation, there are 10 collections of held-out predictions that do not suffer from the severe bias that occurs when we simple re-predict the training set. 
+
+The CV+ estimator ([Barber _et al._ (2021)](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=%22Predictive+inference+with+the+jackknife%2B%22&btnG=)) can be used to assemble a reference distribution of residuals. If we use `fit_resamples()` or one of the `tune_*()` functions, we can gather those residuals and use them to create prediction intervals. We'll have to make sure that we save the out-of-sample predictions as well as the resasmpled models.  
+
+The `control_*()` functions can be used for this purpose. We can set `save_pred = TRUE` to save the predicted values. For the resampled models, we can use the tools to [extract the models](https://tidymodels.github.io/tidymodels_dot_org/learn/index.html#category=extracting%20results) via the `extract` argument. We'll use the `I()` function to return the fitted workflows from each resample: 
+
+```{r}
+#| label: control
+ctrl <- control_resamples(save_pred = TRUE, extract = I)
+```
+
+Let's use 10-fold cross-validation to resample the neural network: 
+
+```{r}
+#| label: cv
+
+set.seed(493)
+folds <- vfold_cv(train_data)
+
+nnet_rs <- 
+  nnet_wflow %>% 
+  fit_resamples(folds, control = ctrl)
+
+collect_metrics(nnet_rs)
+```
+
+The model has an estimated root mean squared error of `r collect_metrics(nnet_rs) %>% dplyr::filter(.metric == "rmse") %>% pluck("mean") %>% signif(3)` and an estimated R<sup>2</sup> of `r collect_metrics(nnet_rs) %>% dplyr::filter(.metric == "rsq") %>% pluck("mean") %>% signif(3)`.
```

Comment:
Is this to get a general sense for how the model is doing? If so, it might be helpful to say that, so that people don't assume that this is necessary for getting prediction intervals.